File: backend/app/schemas/transaction.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import field_validator
from sqlmodel import SQLModel

from .common import TransferDirection, _non_negative


class CashAdjustCreate(SQLModel):
  happened_at: Optional[datetime] = None
  delta_cash: int
  reason: Optional[str] = None
  request_id: Optional[str] = None


class CashAdjustUpdate(SQLModel):
  delta_cash: Optional[int] = None
  reason: Optional[str] = None


class CashAdjustmentRow(SQLModel):
  id: str
  delta_cash: int
  reason: Optional[str] = None
  effective_at: datetime
  created_at: datetime
  is_deleted: bool = False


class ExpenseCategoryCreate(SQLModel):
  name: str


class ExpenseCategoryOut(SQLModel):
  id: str
  name: str
  is_active: bool
  created_at: datetime


class ExpenseCreate(SQLModel):
  happened_at: Optional[datetime] = None
  kind: Literal["expense", "deposit"]
  category_id: Optional[str] = None
  amount: int
  paid_from: Optional[Literal["cash", "bank"]] = None
  note: Optional[str] = None
  vendor: Optional[str] = None
  request_id: Optional[str] = None


class ExpenseOut(SQLModel):
  id: str
  happened_at: datetime
  day: datetime
  kind: str
  category_id: Optional[str] = None
  amount: int
  paid_from: Optional[str] = None
  note: Optional[str] = None
  vendor: Optional[str] = None
  is_reversed: bool
  is_deleted: bool = False


class ExpenseCreateLegacy(SQLModel):
  date: str
  expense_type: str
  amount: int
  note: Optional[str] = None
  request_id: Optional[str] = None
  happened_at: Optional[datetime] = None


class ExpenseUpdate(SQLModel):
  date: Optional[str] = None
  expense_type: Optional[str] = None
  amount: Optional[int] = None
  note: Optional[str] = None
  happened_at: Optional[datetime] = None


class ExpenseOutLegacy(SQLModel):
  id: str
  date: str
  expense_type: str
  amount: int
  note: Optional[str] = None
  created_at: Optional[datetime] = None
  created_by: Optional[str] = None
  is_deleted: bool = False


class CompanyCylinderSettleCreate(SQLModel):
  gas_type: str
  quantity: int
  direction: Literal["receive_full", "return_empty"]
  happened_at: Optional[datetime] = None
  note: Optional[str] = None
  request_id: Optional[str] = None

  @field_validator("quantity")
  @classmethod
  def _validate_non_negative(cls, value: Optional[int], info) -> Optional[int]:
    return _non_negative(value, info.field_name)


class CompanyCylinderSettleOut(SQLModel):
  id: str
  happened_at: datetime
  gas_type: str
  quantity: int
  direction: Literal["receive_full", "return_empty"]
  note: Optional[str] = None


class CompanyPaymentCreate(SQLModel):
  amount: int
  note: Optional[str] = None
  request_id: Optional[str] = None
  happened_at: Optional[datetime] = None
  date: Optional[str] = None
  time: Optional[str] = None
  time_of_day: Optional[Literal["morning", "evening"]] = None
  at: Optional[str] = None


class CompanyPaymentOut(SQLModel):
  id: str
  happened_at: datetime
  amount: int
  note: Optional[str] = None
  is_deleted: bool = False


class CompanyBuyIronCreate(SQLModel):
  new12: int = 0
  new48: int = 0
  total_cost: int = 0
  paid_now: int = 0
  note: Optional[str] = None
  request_id: Optional[str] = None
  happened_at: Optional[datetime] = None
  date: Optional[str] = None
  time: Optional[str] = None
  time_of_day: Optional[Literal["morning", "evening"]] = None
  at: Optional[str] = None

  @field_validator("new12", "new48", "total_cost", "paid_now")
  @classmethod
  def _validate_non_negative(cls, value: Optional[int], info) -> Optional[int]:
    return _non_negative(value, info.field_name)


class CompanyBuyIronOut(SQLModel):
  id: str
  happened_at: datetime
  new12: int
  new48: int
  total_cost: int
  paid_now: int
  note: Optional[str] = None


class CompanyBalanceAdjustmentCreate(SQLModel):
  money_balance: int = 0
  cylinder_balance_12: int = 0
  cylinder_balance_48: int = 0
  note: Optional[str] = None
  request_id: Optional[str] = None
  happened_at: Optional[datetime] = None
  date: Optional[str] = None
  time: Optional[str] = None
  time_of_day: Optional[Literal["morning", "evening"]] = None
  at: Optional[str] = None


class CompanyBalanceAdjustmentOut(SQLModel):
  id: str
  happened_at: datetime
  money_balance: int
  cylinder_balance_12: int
  cylinder_balance_48: int
  note: Optional[str] = None


class CompanyBalancesOut(SQLModel):
  company_money: int
  company_cyl_12: int
  company_cyl_48: int
  inventory_full_12: int
  inventory_empty_12: int
  inventory_full_48: int
  inventory_empty_48: int


class BankDepositCreate(SQLModel):
  happened_at: Optional[datetime] = None
  amount: int
  direction: TransferDirection = "wallet_to_bank"
  note: Optional[str] = None
  request_id: Optional[str] = None

  @field_validator("amount")
  @classmethod
  def _validate_amount(cls, value: Optional[int], info) -> Optional[int]:
    return _non_negative(value, info.field_name)


class BankDepositOut(SQLModel):
  id: str
  happened_at: datetime
  amount: int
  direction: TransferDirection
  note: Optional[str] = None
  is_deleted: bool = False

File: frontend/types/transaction.ts
import { z } from "zod";

export const CashAdjustmentSchema = z
  .object({
    id: z.string(),
    delta_cash: z.number(),
    reason: z.string().nullish(),
    effective_at: z.string(),
    created_at: z.string(),
    is_deleted: z.boolean().optional(),
  })
  .passthrough();
export type CashAdjustment = z.infer<typeof CashAdjustmentSchema>;

export const CashAdjustmentCreateSchema = z.object({
  date: z.string().optional(),
  time: z.string().optional(),
  delta_cash: z.number(),
  reason: z.string().optional(),
  happened_at: z.string().optional(),
});
export type CashAdjustmentCreate = z.infer<typeof CashAdjustmentCreateSchema>;

export const CashAdjustmentUpdateSchema = z.object({
  delta_cash: z.number().optional(),
  reason: z.string().optional(),
});
export type CashAdjustmentUpdate = z.infer<typeof CashAdjustmentUpdateSchema>;

export const CompanyPaymentSchema = z.object({
  id: z.string(),
  happened_at: z.string(),
  created_at: z.string().optional(),
  amount: z.number(),
  note: z.string().nullish(),
  is_deleted: z.boolean().optional(),
  live_debt_cash: z.number().nullish(),
});
export type CompanyPayment = z.infer<typeof CompanyPaymentSchema>;
export type CompanyPaymentCreateInput = {
  amount: number;
  note?: string;
  date?: string;
  time?: string;
  happened_at?: string;
  request_id?: string;
};

export const CompanyBuyIronSchema = z.object({
  id: z.string(),
  happened_at: z.string(),
  new12: z.number(),
  new48: z.number(),
  total_cost: z.number(),
  paid_now: z.number(),
  note: z.string().nullish(),
});
export type CompanyBuyIron = z.infer<typeof CompanyBuyIronSchema>;
export type CompanyBuyIronCreateInput = {
  new12: number;
  new48: number;
  total_cost: number;
  paid_now: number;
  note?: string;
  date?: string;
  time?: string;
  time_of_day?: "morning" | "evening";
  happened_at?: string;
  request_id?: string;
};

export const CompanyBalanceAdjustmentSchema = z.object({
  id: z.string(),
  happened_at: z.string(),
  created_at: z.string().optional(),
  money_balance: z.number(),
  cylinder_balance_12: z.number(),
  cylinder_balance_48: z.number(),
  delta_money: z.number().optional().default(0),
  delta_cylinder_12: z.number().optional().default(0),
  delta_cylinder_48: z.number().optional().default(0),
  live_debt_cash: z.number().nullish(),
  live_debt_cylinders_12: z.number().nullish(),
  live_debt_cylinders_48: z.number().nullish(),
  note: z.string().nullish(),
  is_deleted: z.boolean().optional(),
});
export type CompanyBalanceAdjustment = z.infer<typeof CompanyBalanceAdjustmentSchema>;
export type CompanyBalanceAdjustmentCreateInput = {
  money_balance: number;
  cylinder_balance_12: number;
  cylinder_balance_48: number;
  note?: string;
  date?: string;
  time?: string;
  time_of_day?: "morning" | "evening";
  happened_at?: string;
  request_id?: string;
};
export type CompanyBalanceAdjustmentUpdateInput = {
  money_balance?: number;
  cylinder_balance_12?: number;
  cylinder_balance_48?: number;
  note?: string;
  date?: string;
  time?: string;
  time_of_day?: "morning" | "evening";
  happened_at?: string;
};

export const CompanyBalancesSchema = z.object({
  company_money: z.number(),
  company_cyl_12: z.number(),
  company_cyl_48: z.number(),
  inventory_full_12: z.number(),
  inventory_empty_12: z.number(),
  inventory_full_48: z.number(),
  inventory_empty_48: z.number(),
});
export type CompanyBalances = z.infer<typeof CompanyBalancesSchema>;

export const ExpenseSchema = z
  .object({
    id: z.string(),
    date: z.string(),
    happened_at: z.string().nullish(),
    expense_type: z.string(),
    amount: z.number(),
    note: z.string().nullish(),
    created_at: z.string().optional(),
    created_by: z.string().nullish(),
    is_deleted: z.boolean().optional(),
  })
  .passthrough();
export type Expense = z.infer<typeof ExpenseSchema>;

export const ExpenseCreateInputSchema = z.object({
  date: z.string(),
  time: z.string().optional(),
  expense_type: z.string(),
  amount: z.number(),
  note: z.string().nullish().optional(),
  created_by: z.string().nullish().optional(),
  happened_at: z.string().optional(),
});
export type ExpenseCreateInput = z.infer<typeof ExpenseCreateInputSchema>;

export const ExpenseUpdateInputSchema = ExpenseCreateInputSchema.partial();
export type ExpenseUpdateInput = z.infer<typeof ExpenseUpdateInputSchema>;

export const BankDepositSchema = z.object({
  id: z.string(),
  happened_at: z.string(),
  created_at: z.string().optional(),
  amount: z.number(),
  direction: z.enum(["wallet_to_bank", "bank_to_wallet"]),
  note: z.string().nullish(),
  is_deleted: z.boolean().optional(),
});
export type BankDeposit = z.infer<typeof BankDepositSchema>;
